prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+            n_trials (int): The number of trials for conducting hyperparameter search
+        """
+        self.generation_model_trainer = trainer
+        self.n_trials = n_trials
+
+    def optimize_hyperparameters(
+        self,
+        training_datasets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: Optional[dict[str, Any]] = None,
+    ) -> dict[str, Any]:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_datasets (list[Dataset]): One or more training datasets or
+                subset of the actual dataset for running the trainer.
+            validation_sets (Dataset): A dataset for computing validation metrics.
+            hyperparameter_space (Optional[dict[str, Any]], optional): The set
+                of possible values of hyperparaneters values required for doing
+                optimal hyperparameter search. Defaults to None.
+
+        Returns:
+            Returns a dict which contains the best hyperparameters
+        """
+        supported_hp_space_keys = set(DEFAULT_HYPERPARAMETERS_SPACE.keys())
+        if hyperparameters is not None:
+            assert set(hyperparameters.keys()).issubset(
+                supported_hp_space_keys
+            ), f"Only support {supported_hp_space_keys} as training parameters."
+        hyperparameter_space = self._build_hp_space(hyperparameters)
+
+        concatenated_training_dataset = concatenate_datasets(training_datasets)
+        train_dataset = self.generation_model_trainer.tokenize_dataset(
+            concatenated_training_dataset
+        )
+
+        if isinstance(validation, list):
+            validation = concatenate_datasets(validation)
+            validation = self.generation_model_trainer.tokenize_dataset(validation)
+
+        def objective(trial: Trial):
+            model = self.generation_model_trainer.model
+            training_args = transformers.TrainingArguments(
+                output_dir="./checkpoint",
+                learning_rate=trial.suggest_loguniform(
+                    "learning_rate",
+                    low=hyperparameter_space["min_learning_rate"],
+                    high=hyperparameter_space["max_learning_rate"],
+                ),
+                weight_decay=trial.suggest_loguniform(
+                    "weight_decay",
+                    low=hyperparameter_space["min_weight_decay"],
+                    high=hyperparameter_space["max_weight_decay"],
+                ),
+                num_train_epochs=trial.suggest_int(
+                    "num_train_epochs",
+                    low=hyperparameter_space["min_num_train_epochs"],
+                    high=hyperparameter_space["max_num_train_epochs"],
+                ),
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
```

Comment:
As mentioned in my comment in `prompt2model/utils/config.py`, I don't see how this hyperparameter will affect the trained model. If you agree, let's remove it from the set of hyperparameters tuned here, and instead use a fixed default value in `prompt2model/utils/config.py`

Reply:
Yeah will remove this, after doing a bit of additional research on anything better strategy if I missed.

Reply:
As mentioned before, I don't think this is necessary or useful for hyperparameter search.

Reply:
Sure, removed it.